src/control/zoom/ZoomControl.cpp

```diff
@@ -78,15 +79,26 @@ auto ZoomControl::getVisibleRect() -> Rectangle<double> {
 }
 
 void ZoomControl::setScrollPositionAfterZoom(utl::Point<double> scrollPos) {
-    this->scrollPosition = (scrollPos + this->zoomWidgetPos) / this->zoom;
+    size_t currentPageIdx = this->view->getCurrentPage();
+
+    GtkWidget* widget = view->getWidget();
+    Layout* layout = gtk_xournal_get_layout(widget);
+
+    // Not everything changes size as we zoom in/out. The padding, for example,
+    // remains constant!
+    this->unscaledPixels = {0, static_cast<double>(layout->getPaddingAbovePage(currentPageIdx))};
+
+    // Use this->zoomWidgetPos to zoom into a location other than the top-left (e.g. where
+    // the user pinched).
+    this->scrollPosition = (scrollPos - this->unscaledPixels + this->zoomWidgetPos) / this->zoom;
 }
 
 auto ZoomControl::getScrollPositionAfterZoom() const -> utl::Point<double> {
     if (this->zoomSequenceStart == -1) {
         return {-1, -1};
     }
 
-    return (this->scrollPosition * this->zoom) - this->zoomWidgetPos;
+    return (this->scrollPosition * this->zoom) - this->zoomWidgetPos + this->unscaledPixels;
```

Comment:
Not a very good idea to use the variable for this. This forces the call of `void ZoomControl::setScrollPositionAfterZoom(utl::Point<double> scrollPos)` before any call to `ZoomControl::getScrollPositionAfterZoom()` may happen or it returns the wrong result. As getPaddingAbovePage is also const you should just call the method here as well.

Reply:
Notice that `ZoomControl::startZoomSequence` calls `ZoomControl::setScrollPositionAfterZoom`! 

Of course, I can still make this change -- it gets rid of an unnecessary member variable ☺.

Reply:
`ZoomControl::startZoomSequence` calls `ZoomControl::setScrollPositionAfterZoom`. To call `getPaddingAbovePage`, we also need to call `XournalView::getWidget()`, which is not `const` ☺.

Reply:
I should document this!

Reply:
Done!

I should note that this change is only a partial fix. While the zooming in/out very far into a large document no longer causes the viewport to scroll significantly, it does cause the viewport to rapidly move up and down small amounts.

For anyone working on this in the future: Having set `XOURNAL_PADDING_BETWEEN` and tested before making these changes, I saw that this behavior occurs without this new partial fix.